### PR TITLE
Gateway: add support for HTTP OPTIONS request type

### PR DESCRIPTION
**Pull request by lidel**
This potentially closes https://github.com/ipfs/go-ipfs/issues/934

`OPTIONS` is a noop request that is used by the browsers to check if server explicitly accepts  cross-site `XMLHttpRequest` (XHR).  The acceptance is indicated by the presence of CORS headers.

Before this fix user could enable CORS headers in the Gateway config (note that this disabled by default):

    ipfs config --json Gateway.HTTPHeaders.Access-Control-Allow-Origin '["*"]'
    ipfs config --json Gateway.HTTPHeaders.Access-Control-Allow-Methods '["PUT", "GET", "POST"]'
    ipfs config --json Gateway.HTTPHeaders.Access-Control-Allow-Headers '["X-Requested-With"]'

..but requests sent from JavaScript failed due to the lack of support for `OPTIONS` request type.
I described details of the problem with XHR in https://github.com/lidel/ipfs-firefox-addon/issues/45#issuecomment-172054478.

With this fix sending cross-site XHR to the local Gateway works as expected.

Reference:
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Preflighted_requests
- https://en.wikipedia.org/wiki/Cross-origin_resource_sharing#Preflight_example

### CODE REVIEW

**Comment by lgierth** [comment]
Thanks! -- could you add a test which makes sure this works for both the gateway and the readonly api at `:8080/api`?

**Comment by lidel** [comment]
@lgierth where should I add OPTION tests?
Gateway in  `core/corehttp/gateway_test.go`  and API in `commands/http/handler_test.go` ?

BTW: while OPTIONS itself works with API, CORS headers such as `Access-Control-Allow-*` are removed from `/api/v0/*` responses as a security  measure (even if user explicitly added them):  
- https://github.com/ipfs/go-ipfs/blob/5d9ee59908099df3f7e85679f7384c98d4ac8111/commands/http/handler.go#L71-L82
- https://github.com/ipfs/go-ipfs/pull/1529#issuecomment-125702347

In short: cross-site XHR durrently does not work for API resources.

File: core/corehttp/gateway_handler.go [hunk 1]
```diff
@@ -82,6 +82,11 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if r.Method == "OPTIONS" {
```
Comment (rht) [inline core/corehttp/gateway_handler.go]: Is this enabled in both ro and writable case?

Comment (lidel) [inline core/corehttp/gateway_handler.go]: I feel we should always produce `OPTIONS` response.
Gateway owner can decide what types of requests are accepted by customizing  `Access-Control-Allow-Method` header (`GET` is ro, `POST`, `PUT`, `DELETE` are rw). 
All major browsers respect this header and refuse to do XHR requests if method is not whitelisted.

Comment (rht) [inline core/corehttp/gateway_handler.go]: ic

**Comment by rht** [comment]
@lidel I think the test is to be put in test/sharness/t0110-gateway.sh.

To speed-up this PR, here is a quick test:
```shell
test_expect_success "OPTIONS to gateway succeeds" '
    curl -svX OPTIONS "http://localhost:$port/" 2> curl_output
'
test_expect_success "OPTIONS to gateway output looks good" '
    echo "expected string" >expected &&
    test_cmp expected curl_output
'

#for gateway api
```

**Comment by rht** [comment]
I figured that the cors wrapper in https://github.com/ipfs/go-ipfs/blob/a2b0287a5f6fb5c0f35b63688ad5ac81c451afce/commands/http/handler.go#L102 was meant to handle preflight (see https://github.com/rs/cors/blob/45d446e551b0020074e771dee17d2bb2fd2e9b44/cors.go#L245).

Somehow it doesn't get enabled.

**Comment by rht** [comment]
1. It appears that the `github.com/rs/cors` library used in the current master is outdated. Either 1. use that lib, or 2. replicate the essentially functionality (only show Origin, Access-Control-Request-Method, Access-Control-Request-Headers, instead of `addUserHeaders`).
2. I tested the pr and it didn't work on the ro api. Either leave this to a separate PR, or put a few lines in commands/http/handler.go.

**Comment by rht** [comment]
Just updated the test. It should be runnable.

**Comment by lidel** [comment]
Thank you @rht, I decided to add CORS headers and OPTIONS tests as a separate file:
- I don't want to pollute other gateway tests with CORS headers
- (as of now) changing headers requires daemon restart anyway to see the changes
- we may want to add more CORS-related tests in future as our browser addons and IPFS-based app ecosystem matures

Let me know if I should add anything else for this to be merged.

**Comment by rht** [comment]
There could be test for gateway ACL as well.
The test file LGTM.
The travis-ci build stall is orthogonal to this pr.
This is ready to merge, I think.

**Comment by jbenet** [comment]
LGTM!

**Comment by jbenet** [comment]
thanks @lidel
